flow/layers/physical_shape_layer.cc

```diff
@@ -82,6 +93,13 @@ void PhysicalShapeLayer::Paint(PaintContext& context) const {
   TRACE_EVENT0("flutter", "PhysicalShapeLayer::Paint");
   FML_DCHECK(needs_painting());
 
+#if defined(OS_FUCHSIA)
+  if (fuchsia_painted_) {
```

Comment:
what is the difference between fuchsia_painted_ and needs_painting() ? When/why do their values diverge?

Reply:
needs_painting check whether the paint_bounds is empty and it applies to all Flutter platforms.

fuchsia_painted is for Fuchsia only because Fuchsia always allocates a buffer for a PhysicalShapeLayer. Even if the paint_bounds is not empty, we can still use this to optimize performance by only paint the same PhysicalShapeLayer once within Fuchsia.